File: src/accessiweather/noaa_radio/session.py
"""Shared NOAA Weather Radio playback session."""

from __future__ import annotations

from collections.abc import Callable

from accessiweather.noaa_radio.player import RadioPlayer
from accessiweather.noaa_radio.stations import Station


class RadioSession:
    """Owns radio stream state independently from any one dialog."""

    def __init__(self) -> None:
        """Initialize the shared playback session."""
        self.playing_station: Station | None = None
        self.current_urls: list[str] = []
        self.current_url_index = 0
        self._on_playing: Callable[[], None] | None = None
        self._on_stopped: Callable[[], None] | None = None
        self._on_error: Callable[[str], None] | None = None
        self._on_stalled: Callable[[], None] | None = None
        self._on_reconnecting: Callable[[int], None] | None = None
        self.player = RadioPlayer(
            on_playing=self._handle_playing,
            on_stopped=self._handle_stopped,
            on_error=self._handle_error,
            on_stalled=self._handle_stalled,
            on_reconnecting=self._handle_reconnecting,
        )

    def bind_callbacks(
        self,
        *,
        on_playing: Callable[[], None],
        on_stopped: Callable[[], None],
        on_error: Callable[[str], None],
        on_stalled: Callable[[], None],
        on_reconnecting: Callable[[int], None],
    ) -> None:
        """Attach UI callbacks for the currently visible dialog."""
        self._on_playing = on_playing
        self._on_stopped = on_stopped
        self._on_error = on_error
        self._on_stalled = on_stalled
        self._on_reconnecting = on_reconnecting

    def unbind_callbacks(self) -> None:
        """Detach UI callbacks when the dialog is dismissed."""
        self._on_playing = None
        self._on_stopped = None
        self._on_error = None
        self._on_stalled = None
        self._on_reconnecting = None

    def is_playing(self) -> bool:
        """Return whether a stream is currently active."""
        return self.player.is_playing()

    def stop(self, *, notify: bool = True) -> None:
        """Stop playback and clear session state."""
        self.player.stop(notify=notify)
        self.playing_station = None

    def _handle_playing(self) -> None:
        if self._on_playing is not None:
            self._on_playing()

    def _handle_stopped(self) -> None:
        self.playing_station = None
        if self._on_stopped is not None:
            self._on_stopped()

    def _handle_error(self, message: str) -> None:
        self.playing_station = None
        if self._on_error is not None:
            self._on_error(message)

    def _handle_stalled(self) -> None:
        if self._on_stalled is not None:
            self._on_stalled()

    def _handle_reconnecting(self, attempt: int) -> None:
        if self._on_reconnecting is not None:
            self._on_reconnecting(attempt)


_shared_session: RadioSession | None = None


def get_shared_radio_session() -> RadioSession:
    """Return the app-wide NOAA radio playback session."""
    global _shared_session
    if _shared_session is None:
        _shared_session = RadioSession()
    return _shared_session


def stop_shared_radio_session() -> None:
    """Stop the app-wide NOAA radio session, if it exists."""
    if _shared_session is not None:
        _shared_session.stop()

File: rust/crates/aw-radio/src/session.rs
//! The app-wide NOAA Weather Radio playback session, shared by the dialog,
//! the global hotkey and alert auto-tune. It outlives any one dialog.
//!
//! Ports `noaa_radio/session.py`. Python keeps one module-level session
//! (`get_shared_radio_session`); the Rust app creates one `Arc<RadioSession>`
//! (see [`crate::RadioContext`]).

use std::sync::{Arc, Mutex, Weak};

use crate::player::{AudioBackend, PlayerEvent, RadioPlayer};
use crate::preferences::SharedPreferences;
use crate::stations::Station;

/// Session callback for whoever is showing playback (the open dialog).
/// Called on the thread that produced the event; the receiver is
/// responsible for getting onto its UI thread.
pub type SessionCallback = Arc<dyn Fn(PlayerEvent) + Send + Sync>;

/// What is on the air.
#[derive(Debug, Clone, Default, PartialEq)]
pub struct SessionState {
    pub playing_station: Option<Station>,
    pub current_urls: Vec<String>,
    pub current_url_index: usize,
}

pub struct RadioSession {
    pub player: RadioPlayer,
    preferences: Option<SharedPreferences>,
    state: Mutex<SessionState>,
    callback: Mutex<Option<SessionCallback>>,
}

impl RadioSession {
    pub fn new(
        backend: Arc<dyn AudioBackend>,
        preferences: Option<SharedPreferences>,
    ) -> Arc<Self> {
        Arc::new_cyclic(|weak: &Weak<RadioSession>| {
            let weak = weak.clone();
            let on_event = Arc::new(move |event: PlayerEvent| {
                if let Some(session) = weak.upgrade() {
                    session.handle_player_event(event);
                }
            });
            Self {
                player: RadioPlayer::new(backend, on_event),
                preferences,
                state: Mutex::default(),
                callback: Mutex::new(None),
            }
        })
    }

    /// Attach the visible dialog's callback.
    pub fn bind_callbacks(&self, callback: SessionCallback) {
        *self.callback.lock().unwrap() = Some(callback);
    }

    /// Detach it when the dialog closes (playback keeps going).
    pub fn unbind_callbacks(&self) {
        *self.callback.lock().unwrap() = None;
    }

    pub fn is_playing(&self) -> bool {
        self.player.is_playing()
    }

    /// Stop playback and forget the station.
    pub fn stop(&self, notify: bool) {
        self.player.stop(notify);
        self.state.lock().unwrap().playing_station = None;
    }

    pub fn state(&self) -> SessionState {
        self.state.lock().unwrap().clone()
    }

    pub fn playing_station(&self) -> Option<Station> {
        self.state.lock().unwrap().playing_station.clone()
    }

    /// Change session state atomically. Do not call player methods inside.
    pub fn update<R>(&self, change: impl FnOnce(&mut SessionState) -> R) -> R {
        change(&mut self.state.lock().unwrap())
    }

    fn handle_player_event(&self, event: PlayerEvent) {
        match &event {
            // Both the dialog and auto-tune set the station before starting
            // the player, so recording it here covers every path on the air.
            PlayerEvent::Playing => self.remember_last_station(),
            PlayerEvent::Stopped | PlayerEvent::Error(_) => {
                self.state.lock().unwrap().playing_station = None;
            }
            PlayerEvent::Stalled | PlayerEvent::Reconnecting(_) => {}
        }
        let callback = self.callback.lock().unwrap().clone();
        if let Some(callback) = callback {
            callback(event);
        }
    }

    /// Persist the station so the hotkey can resume it later.
    fn remember_last_station(&self) {
        let (Some(prefs), Some(station)) = (&self.preferences, self.playing_station()) else {
            return;
        };
        prefs
            .lock()
            .unwrap()
            .set_last_station(Some(&station.call_sign));
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::player::fake::FakeBackend;
    use crate::preferences::RadioPreferences;

    fn wxk27() -> Station {
        Station::new("WXK27", 162.4, "Austin", 30.2672, -97.7431, "TX")
    }

    #[test]
    fn records_last_station_and_forwards_events() {
        let prefs = RadioPreferences::new(None).shared();
        let session = RadioSession::new(Arc::new(FakeBackend::default()), Some(prefs.clone()));
        let events = Arc::new(Mutex::new(Vec::new()));
        let sink = events.clone();
        session.bind_callbacks(Arc::new(move |e| sink.lock().unwrap().push(e)));

        // No station: nothing recorded.
        assert!(session.player.play("http://a"));
        assert_eq!(prefs.lock().unwrap().get_last_station(), None);

        // Callers switching streams stop quietly first, as the dialog does.
        session.player.stop(false);
        session.update(|s| s.playing_station = Some(wxk27()));
        assert!(session.player.play("http://b"));
        assert_eq!(
            prefs.lock().unwrap().get_last_station().as_deref(),
            Some("WXK27")
        );
        assert_eq!(session.playing_station(), Some(wxk27()));
        assert!(session.is_playing());

        session.stop(true);
        assert_eq!(session.playing_station(), None);
        assert!(!session.is_playing());
        assert_eq!(
            *events.lock().unwrap(),
            [
                PlayerEvent::Playing,
                PlayerEvent::Playing,
                PlayerEvent::Stopped
            ]
        );
    }

    #[test]
    fn error_clears_station_and_unbind_keeps_playing() {
        let session = RadioSession::new(FakeBackend::with_results(&[false, true]), None);
        session.update(|s| s.playing_station = Some(wxk27()));
        assert!(!session.player.play("http://a"));
        assert_eq!(session.playing_station(), None);

        let events = Arc::new(Mutex::new(0));
        let sink = events.clone();
        session.bind_callbacks(Arc::new(move |_| *sink.lock().unwrap() += 1));
        session.unbind_callbacks();
        assert!(session.player.play("http://b"));
        assert!(session.is_playing());
        assert_eq!(*events.lock().unwrap(), 0);
    }
}
